-1

**Failed Tests**:   Build 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bbaa8e/18616/summary.html
**COMMIT**: 3c011ebbd6642058e680c6b3a8d0304d451c0caa
**CMSSW**: CMSSW_12_1_X_2021-09-14-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35276/18616/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_1_X_2021-09-14-1100/src/RecoJets/FFTJetProducers/plugins/FFTJetProducer.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_1_X_2021-09-14-1100/src/RecoJets/FFTJetProducers/plugins/FFTJetTreeDump.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_1_X_2021-09-14-1100/src/RecoJets/FFTJetProducers/plugins/FFTJetVertexAdder.cc
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_1_X_2021-09-14-1100/src/RecoJets/FFTJetProducers/plugins/FFTJetProducer.cc: In instantiation of 'void FFTJetProducer::writeJets(edm::Event&, const edm::EventSetup&) [with Jet = reco::CaloJet]':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_1_X_2021-09-14-1100/src/RecoJets/FFTJetProducers/plugins/FFTJetProducer.cc:604:3:   required from here
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_1_X_2021-09-14-1100/src/RecoJets/FFTJetProducers/plugins/FFTJetProducer.cc:568:18: error: no matching function for call to 'writeSpecific(reco::CaloJet&, fftjetcms::VectorLike&, const Point&, __gnu_cxx::__alloc_traits<std::allocator<std::vector<edm::Ptr<reco::Candidate> > >, std::vector<edm::Ptr<reco::Candidate> > >::value_type&, const edm::EventSetup&)'
  568 |     writeSpecific(jet, jet4vec, vertexUsed(), constituents[ijet + 1], iSetup);
      |     ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_1_X_2021-09-14-1100/src/RecoJets/FFTJetProducers/plugins/FFTJetProducer.cc:40:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_1_X_2021-09-14-1100/src/RecoJets/JetProducers/interface/JetSpecific.h:28:8: note: candidate: 'void reco::writeSpecific(reco::CaloJet&, const LorentzVector&, const Point&, const std::vector<edm::Ptr<reco::Candidate> >&, const CaloGeometry&, const HcalTopology&)'
   28 |   void writeSpecific(reco::CaloJet& jet,
</pre>

--- a/RecoJets/FFTJetProducers/plugins/FFTJetProducer.cc
+++ b/RecoJets/FFTJetProducers/plugins/FFTJetProducer.cc
@@ -161,6 +161,11 @@ FFTJetProducer::FFTJetProducer(const edm::ParameterSet& ps)
   const std::string alias(ps.getUntrackedParameter<std::string>("alias", outputLabel));
   jet_type_switch(makeProduces, alias, outputLabel);
 
+  if (jetType == CALOJET) {
+    geometry_token_ = esConsumes();
+    topology_token_ = esConsumes();
+  }
+
   // Build the set of pattern recognition scales.
   // This is needed in order to read the clustering tree
   // from the event record.
@@ -568,11 +573,9 @@ void FFTJetProducer::writeJets(edm::Event& iEvent, const edm::EventSetup& iSetup
     // call the appropriate specific code.
     T jet;
     if constexpr (std::is_same_v<T, reco::CaloJet>) {
-      edm::ESHandle<CaloGeometry> geometry;
-      iSetup.get<CaloGeometryRecord>().get(geometry);
-      edm::ESHandle<HcalTopology> topology;
-      iSetup.get<HcalRecNumberingRecord>().get(topology);
-      writeSpecific(jet, jet4vec, vertexUsed(), constituents[ijet + 1], *geometry, *topology);
+      const CaloGeometry& geometry = iSetup.getData(geometry_token_);
+      const HcalTopology& topology = iSetup.getData(topology_token_);
+      writeSpecific(jet, jet4vec, vertexUsed(), constituents[ijet + 1], geometry, topology);
     } else {
       writeSpecific(jet, jet4vec, vertexUsed(), constituents[ijet + 1]);
     }

--- a/RecoJets/FFTJetProducers/plugins/FFTJetProducer.cc
+++ b/RecoJets/FFTJetProducers/plugins/FFTJetProducer.cc
@@ -38,6 +38,8 @@
 #include "DataFormats/Candidate/interface/Candidate.h"
 
 #include "RecoJets/JetProducers/interface/JetSpecific.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+#include "Geometry/Records/interface/HcalRecNumberingRecord.h"
 
 // Loader for the lookup tables
 #include "JetMETCorrections/FFTJetObjects/interface/FFTJetLookupTableSequenceLoader.h"
@@ -565,7 +567,15 @@ void FFTJetProducer::writeJets(edm::Event& iEvent, const edm::EventSetup& iSetup
     // vertex, constituents). These are overridden functions that will
     // call the appropriate specific code.
     T jet;
-    writeSpecific(jet, jet4vec, vertexUsed(), constituents[ijet + 1], iSetup);
+    if constexpr (std::is_same_v<T, reco::CaloJet>) {
+      edm::ESHandle<CaloGeometry> geometry;
+      iSetup.get<CaloGeometryRecord>().get(geometry);
+      edm::ESHandle<HcalTopology> topology;
+      iSetup.get<HcalRecNumberingRecord>().get(topology);
+      writeSpecific(jet, jet4vec, vertexUsed(), constituents[ijet + 1], *geometry, *topology);
+    } else {
+      writeSpecific(jet, jet4vec, vertexUsed(), constituents[ijet + 1]);
+    }
 
     // calcuate the jet area
     double ncells = myjet.ncells();

--- a/RecoJets/FFTJetProducers/plugins/FFTJetProducer.h
+++ b/RecoJets/FFTJetProducers/plugins/FFTJetProducer.h
@@ -71,6 +71,11 @@ namespace fftjetcms {
   class DiscretizedEnergyFlow;
 }
 
+class CaloGeometry;
+class CaloGeometryRecord;
+class HcalTopology;
+class HcalRecNumberingRecord;
+
 //
 // class declaration
 //
@@ -398,6 +403,9 @@ private:
   edm::EDGetTokenT<std::vector<reco::FFTAnyJet<reco::GenJet> > > input_genjet_token_;
   edm::EDGetTokenT<reco::DiscretizedEnergyFlow> input_energyflow_token_;
   edm::EDGetTokenT<reco::FFTJetPileupSummary> input_pusummary_token_;
+
+  edm::ESGetToken<CaloGeometry, CaloGeometryRecord> geometry_token_;
+  edm::ESGetToken<HcalTopology, HcalRecNumberingRecord> topology_token_;
 };
 
 #endif  // RecoJets_FFTJetProducers_FFTJetProducer_h

--- a/RecoJets/JetProducers/src/JetSpecific.cc
+++ b/RecoJets/JetProducers/src/JetSpecific.cc
@@ -35,13 +35,7 @@ void reco::writeSpecific(reco::CaloJet& jet,
                          std::vector<reco::CandidatePtr> const& constituents,
                          CaloGeometry const& geometry,
                          HcalTopology const& topology) {
-  // Get geometry
-  //edm::ESHandle<CaloGeometry> geometry;
-  //c.get<CaloGeometryRecord>().get(geometry);
   const CaloSubdetectorGeometry* towerGeometry = geometry.getSubdetectorGeometry(DetId::Calo, CaloTowerDetId::SubdetId);
-
-  //edm::ESHandle<HcalTopology> topology;
-  //c.get<HcalRecNumberingRecord>().get(topology);
 
   // Make the specific
   reco::CaloJet::Specific specific;

--- a/TopQuarkAnalysis/TopEventProducers/src/PseudoTopProducer.cc
+++ b/TopQuarkAnalysis/TopEventProducers/src/PseudoTopProducer.cc
@@ -150,7 +150,7 @@ void PseudoTopProducer::produce(edm::Event& event, const edm::EventSetup& eventS
 
     const LorentzVector jetP4(fjJet.px(), fjJet.py(), fjJet.pz(), fjJet.E());
     reco::GenJet lepJet;
-    reco::writeSpecific(lepJet, jetP4, genVertex_, constituents, eventSetup);
+    reco::writeSpecific(lepJet, jetP4, genVertex_, constituents);
 
     lepJet.setPdgId(lepCand->pdgId());
     lepJet.setCharge(lepCand->charge());
@@ -226,7 +226,7 @@ void PseudoTopProducer::produce(edm::Event& event, const edm::EventSetup& eventS
 
     const LorentzVector jetP4(fjJet.px(), fjJet.py(), fjJet.pz(), fjJet.E());
     reco::GenJet genJet;
-    reco::writeSpecific(genJet, jetP4, genVertex_, constituents, eventSetup);
+    reco::writeSpecific(genJet, jetP4, genVertex_, constituents);
 
     const double jetArea = fjJet.has_area() ? fjJet.area() : 0;
     genJet.setJetArea(jetArea);

